### DISCUSSION TITLE
Renovate messes up code due to doing an `s/x/y/g`-style substitution

### DISCUSSION THREAD

**Discussion by TWiStErRob** [renovatebot/renovate#23288] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

TL;DR: while renovate does an upgrade on a regexManager dependency, it does a "`replaceAll`" on the whole line rather than replacing the `(?<currentValue>...)` group. This breaks the AGP version in these lines, and correctly upgrades Gradle versions (after the word "on"):
![image](https://github.com/renovatebot/renovate/assets/2906988/d988c8b2-0015-48b5-a68e-90c12046e4bf)

Full details below.

Affected PR: https://github.com/TWiStErRob/net.twisterrob.gradle/pull/544
Commit in question: https://github.com/TWiStErRob/net.twisterrob.gradle/pull/544/commits/a4f4f35c9a86cdb9fa1d97747fcdda4793239e4e
(Recording this here, because it might get force-pushed, and I'll fix the problem manually and merge too.)
Logs: https://developer.mend.io/github/TWiStErRob/net.twisterrob.gradle/-/job/018940f6-19cf-75d6-9ef0-0298fe77eaf6

---

I've observed this multiple times now and happens more often because version numbers align. I'll provide a real life example for now pointing out relevant configuration details. If this is not enough to triage, I can try to create a repro repo.

[Renovate config](https://github.com/TWiStErRob/net.twisterrob.gradle/blob/424183e577805436f5ff42780adf9634cf9cd678/.github/renovate.json#L139-L149)
```json
	"regexManagers": [
		{
			"description": "Update Gradle compatibility CI matrix values and artifact names.",
			"fileMatch": ["^\\.github/workflows/CI\\.yml$"],
			"datasourceTemplate": "gradle-version",
			"depNameTemplate": "gradle",
			"matchStrings": [
				"  gradle: '(?<currentValue>.*)'",
				"AGP (?:.*) on Gradle (?:.*)(?: - plugin)? \\((?:.*) on (?<currentValue>.*)\\) Test Results (?:XMLs|HTMLs|HTML)"
			],
			"versioningTemplate": "gradle"
		},
	],
```
Note: `(?:.*)` is essentially equivalent to `.*` so there's only one capturing group in this regex and it has a name `currentValue`.

Here is an example this matches:
The first of `matchStrings` matches `gradle: '8.2'` in [this GitHub Actions workflow's matrix definition](https://github.com/TWiStErRob/net.twisterrob.gradle/blob/424183e577805436f5ff42780adf9634cf9cd678/.github/workflows/CI.yml#L650-L656), but this part can be ignored.
The second of `matchStrings` matches 3 long lines in the [artifact processing step](https://github.com/TWiStErRob/net.twisterrob.gradle/blob/424183e577805436f5ff42780adf9634cf9cd678/.github/workflows/CI.yml#L968-L974):
```yaml
      - name: "Download 'AGP 8.2.x on Gradle 8.x - plugin (8.2.0-alpha11 on 8.2) Test Results XMLs' artifact."
        if: success() || failure()
        uses: actions/download-artifact@v3
        with:
          name: 'AGP 8.2.x on Gradle 8.x - plugin (8.2.0-alpha11 on 8.2) Test Results XMLs'
          path: 'artifacts/AGP 8.2.x on Gradle 8.x - plugin (8.2.0-alpha11 on 8.2) Test Results XMLs'
```
Here's a visualization of the 3 matches: https://regex101.com/r/mtiR94/1
![image](https://github.com/renovatebot/renovate/assets/2906988/face144d-b1cb-4c5d-8f51-4fa30054c434)

Now, today Gradle released 8.2.1, and Renovate opened a PR from Gradle 8.2 to 8.2.1. Notice that the AGP version in the text also happens to be `8.2.x` or `8.2.0-alpha11`. This is important. Here's the diff Renovate generated:
![image](https://github.com/renovatebot/renovate/assets/2906988/1d9a7cb4-f078-4b38-ba81-23644019272a)
notice how it essentially executed a `match.replaceAll("8.2", "8.2.1")` yielding an invalid AGP version number like `8.2.1.0-alpha11` while also correctly upgrading the Gradle version number (after the "on" word).

### Relevant debug logs

<details><summary>Logs</summary>

Let me know what you need. I guess one of many of these would be relevant:
```
DEBUG: Found match at index 26408 (branch="renovate/gradle-8.2.x")
{
  "packageFile": ".github/workflows/CI.yml"
  "depName": "gradle"
}

DEBUG: Value is not updated (branch="renovate/gradle-8.2.x")
{
  "manager": "regex"
  "packageFile": ".github/workflows/CI.yml"
  "expectedValue": "8.2.1"
  "foundValue": "8.2"
}
```
but not saying much.

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by viceice** [comment]
you probably need to configure auto replace template

https://docs.renovatebot.com/configuration-options/#autoreplacestringtemplate

**Reply by TWiStErRob** [reply]
This was my try https://github.com/TWiStErRob/net.twisterrob.gradle/commit/09e7c107adf1035e80a20341c92eb4d96485f0d7, but now all of them are gone: https://github.com/TWiStErRob/net.twisterrob.gradle/pull/544/commits/d43bb26a607e085d2eb157a6f9e075844ffca39f#diff-3ab46ee209a127470fce3c2cf106b1a1dbadbb929a4b5b13656a4bc4ce19c0b8

<details><summary>patch that didn't work</summary>

```patch
Index: .github/renovate.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.github/renovate.json b/.github/renovate.json
--- a/.github/renovate.json	(revision 9d54d1801f0c067dc37f4315482e2eeea5aedc62)
+++ b/.github/renovate.json	(revision 09e7c107adf1035e80a20341c92eb4d96485f0d7)
@@ -103,15 +103,26 @@
 	],
 	"regexManagers": [
 		{
-			"description": "Update AGP compatibility CI matrix values and artifact names.",
+			"description": "Update AGP compatibility CI matrix values.",
+			"fileMatch": ["^\\.github/workflows/CI\\.yml$"],
+			"datasourceTemplate": "maven",
+			"registryUrlTemplate": "https://maven.google.com",
+			"depNameTemplate": "com.android.tools.build:gradle",
+			"matchStrings": [
+				"  agp: '(?<currentValue>.*)'"
+			],
+			"versioningTemplate": "gradle"
+		},
+		{
+			"description": "Update AGP compatibility CI artifact names.",
 			"fileMatch": ["^\\.github/workflows/CI\\.yml$"],
 			"datasourceTemplate": "maven",
 			"registryUrlTemplate": "https://maven.google.com",
 			"depNameTemplate": "com.android.tools.build:gradle",
 			"matchStrings": [
-				"  agp: '(?<currentValue>.*)'",
-				"AGP (?:.*) on Gradle (?:.*)(?: - plugin)? \\((?<currentValue>.*) on (?:.*)\\) Test Results (?:XMLs|HTMLs|HTML)"
+				"AGP (?<agpName>.*) on Gradle (?<gradleName>.*)(?<plugin> - plugin)? \\((?<currentValue>.*) on (?<gradleExact>.*)\\) Test Results (?<artifactType>XMLs|HTMLs|HTML)"
 			],
+			"autoReplaceStringTemplate": "AGP {{{agpName}}} on Gradle {{{gradleName}}}{{{plugin}}} ({{{newValue}}} on {{{gradleExact}}}) Test Results {{{artifactType}}}",
 			"versioningTemplate": "gradle"
 		},
 		{
@@ -137,14 +148,24 @@
 			"versioningTemplate": "gradle"
 		},
 		{
-			"description": "Update Gradle compatibility CI matrix values and artifact names.",
+			"description": "Update Gradle compatibility CI matrix values.",
 			"fileMatch": ["^\\.github/workflows/CI\\.yml$"],
 			"datasourceTemplate": "gradle-version",
 			"depNameTemplate": "gradle",
 			"matchStrings": [
-				"  gradle: '(?<currentValue>.*)'",
-				"AGP (?:.*) on Gradle (?:.*)(?: - plugin)? \\((?:.*) on (?<currentValue>.*)\\) Test Results (?:XMLs|HTMLs|HTML)"
+				"  gradle: '(?<currentValue>.*)'"
 			],
+			"versioningTemplate": "gradle"
+		},
+		{
+			"description": "Update Gradle compatibility CI artifact names.",
+			"fileMatch": ["^\\.github/workflows/CI\\.yml$"],
+			"datasourceTemplate": "gradle-version",
+			"depNameTemplate": "gradle",
+			"matchStrings": [
+				"AGP (?<agpName>.*) on Gradle (?<gradleName>.*)(?<plugin> - plugin)? \\((<agpExact>.*) on (?<currentValue>.*)\\) Test Results (?<artifactType>XMLs|HTMLs|HTML)"
+			],
+			"autoReplaceStringTemplate": "AGP {{{agpName}}} on Gradle {{{gradleName}}}{{{plugin}}} ({{{agpExact}}} on {{{newValue}}}) Test Results {{{artifactType}}}",
 			"versioningTemplate": "gradle"
 		},
 		{

```

</details>


How can I reference arbitrary parts of the match in the `autoReplaceStringTemplate`? I would be happy with `{{{prefix}}}{{{currentValue}}}{{{suffix}}}` as well.

**Reply by TWiStErRob** [reply]
Made a [repro](https://github.com/TWiStErRob/renovate-23288), trying to figure out how to use `autoReplaceStringTemplate`

Edit: repo is deleted, you can find an archive here: https://github.com/TWiStErRob/repros/commit/d19dbe2197eed716be25816610ff85f45a1e46b1

**Reply by TWiStErRob** [reply]
I found that at my first try the regex was incorrect, after fixing I saw `replaceString`s in the logs that were correct, but I'm getting an error: https://github.com/renovatebot/renovate/discussions/23331

**Comment by TWiStErRob** [comment, marked as answer]
recursive regex strategy can also work

https://docs.renovatebot.com/configuration-options/#matchstringsstrategy

For example I narrowed down the match with it:
```json
"regexManagers": [
	{
		"matchStringsStrategy": "any", // default
		"matchStrings": [
			"AGP (?:.*) on Gradle (?:.*)(?: - plugin)? \\((?:.*) on (?<currentValue>.*)\\) Test Results (?:XMLs|HTMLs|HTML)"
		],
	}
]
```
to
```json
"regexManagers": [
	{
		"matchStringsStrategy": "recursive",
		"matchStrings": [
			"AGP (?:.*) on Gradle (?:.*)(?: - plugin)? \\((?:.*) on (?:.*)\\) Test Results (?:XMLs|HTMLs|HTML)",
			"\\(.*\\)",
			"on (?<currentValue>.*?)\\)"
		],
	}
]
```
which means it'll only match
```
name: 'AGP 8.2.x on Gradle 8.x - plugin (8.2.0-alpha11 on 8.2) Test Results XMLs'
                                                       ^^^^^^^
```
where it's safe to blanket replace 8.2 with 8.2.1.

**Comment by TWiStErRob** [comment]
Similar to [matchStringStrategy](https://github.com/renovatebot/renovate/discussions/23288#discussioncomment-6425802), using regex look-around would also work to narrow down the match only to what needs changing:
https://regex101.com/r/NFhILR/1
![image](https://github.com/renovatebot/renovate/assets/2906988/7e720fdc-59a3-4cb3-a0fc-a8fe24839b98)

but it doesn't work in `re2` engine (due to `A quantifier inside a lookbehind makes it non-fixed width`), so this solution might only work on self-hosted environments with RE2 turned off:

https://github.com/renovatebot/renovate/blob/e23e70d7cccf493b4ad0dbea356e705b7c453ddd/lib/util/regex.ts#L11-L22
